### DISCUSSION TITLE
使用 request 在function 中 return 值  (callback function)

### DISCUSSION THREAD

**Discussion by kevindtrt** [Lidemy/forum#143] [其他問題] [answered]
我使用 query string 向網頁讀取資料
使用 request 取回 body，想把body 的值拿來做其他處理。

_guess function 內的 body 想要回傳出來利用，但是印出值的時候都出現 undefined ，請問有可能是哪個環節有問題才會變成 undefined 。_

先感謝大家解惑
(解決問題後我再把 url內容修掉)

```=javascript
const request = require('request')

const url = 'https://lidemy-http-challenge.herokuapp.com/lv10?token=&num='

console.log(typeof guess(1111))

function guess(str) {
  request(
    url+str,
  (err, response, body)  => {
    let ans = body
    console.log('this is ans =',ans)
    return ans
  }
  )

} 

```

**Comment by aszx87410** [comment, marked as answer]
request 的第二個參數是一個 callback 「function」
所以你 return ans，是那個 callback function 的 return，而不是 guess function 的 return

大概就像是這樣，加在 callback 就是 callback 的回傳值，不是 guess 的：

``` js
function guess(str) {
  function callback() {
    return 123
  }
  return undefined
}
```

因此如果你 guess function 想要 return 東西，一定要在那個 callback function 外面 return 才有用，不然永遠都沒辦法把值傳出去
這時候你就會問了：「可是我想要的東西在 callback function 裡面才拿得到，那怎麼辦呢？」
你可能會想說好吧那就這樣寫：

``` js
function guess(str) {
  let ans = null
  request(url+str, (err, response, body) => {
    ans = body
    console.log('this is ans =',ans)
  })
  return ans
} 
```

這樣不就可以正確把值回傳了嗎？

這是標準錯誤答案，因為沒有考量到順序性的問題，實際執行的順序是跑完 `request(url+str, ...)` 那行之後，就發送 request 了
然後就直接執行下一行 `return ans`
等到 response 回來時才去執行那個 callback function
因此這樣寫的話，回傳的東西還是會是 null

那到底該怎麼辦呢？或許你會想起我說過的一句話：「非同步需要回傳值的時候，一定得用 callback」
因此答案是這樣的：

``` js
function guess(str, cb) {
  request(url+str, (err, response, body) => {
    cb(body)
  })
}

guess(1111, ans => {
  console.log(ans)
}) 
```

一定得透過 callback function 把值傳出去，否則做不到

**Comment by kevindtrt** [comment]
雖然先前跟著直播檢討就知道 callback function 這個東西，也先找了一些資料。還沒有花時間研究，就只能抱著似懂非懂閃過，沒料到被題目攔截下來。大概理解您說明的意思，我先多練習幾次熟悉他，感謝您的答覆
